pkg/cmd/codespace/ports_test.go

```diff
@@ -0,0 +1,234 @@
+	portVisibilities := []portVisibility{
+		{
+			number:     80,
+			visibility: "org",
+		},
+		{
+			number:     9999,
+			visibility: "public",
+		},
+	}
+
+	eventResponses := []string{
+		"serverSharing.sharingSucceeded",
+		"serverSharing.sharingSucceeded",
+	}
+
+	portsData := []liveshare.PortNotification{
+		{
+			Success: true,
+			PortUpdate: liveshare.PortUpdate{
+				Port:       80,
+				ChangeKind: liveshare.PortChangeKindUpdate,
+			},
+		},
+		{
+			Success: true,
+			PortUpdate: liveshare.PortUpdate{
+				Port:       9999,
+				ChangeKind: liveshare.PortChangeKindUpdate,
+			},
+		},
+	}
+
+	err := runUpdateVisibilityTest(portVisibilities, eventResponses, portsData)
+
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+}
+
+func TestPortsUpdateVisibilityFailure403(t *testing.T) {
+	portVisibilities := []portVisibility{
+		{
+			number:     80,
+			visibility: "org",
+		},
+		{
+			number:     9999,
+			visibility: "public",
+		},
+	}
+
+	eventResponses := []string{
+		"serverSharing.sharingSucceeded",
+		"serverSharing.sharingFailed",
+	}
+
+	portsData := []liveshare.PortNotification{
+		{
+			Success: true,
+			PortUpdate: liveshare.PortUpdate{
+				Port:       80,
+				ChangeKind: liveshare.PortChangeKindUpdate,
+			},
+		},
+		{
+			Success: false,
+			PortUpdate: liveshare.PortUpdate{
+				Port:        9999,
+				ChangeKind:  liveshare.PortChangeKindUpdate,
+				ErrorDetail: "test error",
+				StatusCode:  403,
+			},
+		},
+	}
+
+	err := runUpdateVisibilityTest(portVisibilities, eventResponses, portsData)
+	if err == nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	if errors.Unwrap(err) != errUpdatePortVisibilityForbidden {
+		t.Errorf("expected: %v, got: %v", errUpdatePortVisibilityForbidden, errors.Unwrap(err))
+	}
+}
+
+func TestPortsUpdateVisibilityFailure(t *testing.T) {
+	portVisibilities := []portVisibility{
+		{
+			number:     80,
+			visibility: "org",
+		},
+		{
+			number:     9999,
+			visibility: "public",
+		},
+	}
+
+	eventResponses := []string{
+		"serverSharing.sharingSucceeded",
+		"serverSharing.sharingFailed",
+	}
+
+	portsData := []liveshare.PortNotification{
+		{
+			Success: true,
+			PortUpdate: liveshare.PortUpdate{
+				Port:       80,
+				ChangeKind: liveshare.PortChangeKindUpdate,
+			},
+		},
+		{
+			Success: false,
+			PortUpdate: liveshare.PortUpdate{
+				Port:        9999,
+				ChangeKind:  liveshare.PortChangeKindUpdate,
+				ErrorDetail: "test error",
+			},
+		},
+	}
+
+	err := runUpdateVisibilityTest(portVisibilities, eventResponses, portsData)
+	if err == nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	var expectedErr *ErrUpdatingPortVisibility
+	if !errors.As(err, &expectedErr) {
+		t.Errorf("expected: %v, got: %v", expectedErr, err)
+	}
+}
+
+type joinWorkspaceResult struct {
+	SessionNumber int `json:"sessionNumber"`
+}
+
+func runUpdateVisibilityTest(portVisibilities []portVisibility, eventResponses []string, portsData []liveshare.PortNotification) error {
```

Comment:
id suggest taking `t *testing.T` as an argument here and calling `t.Helper()` https://pkg.go.dev/testing#T.Helper so that this function is not in the stack trace if there's an error since its not actually the thing being tested

Reply:
done